File: app/src/main/java/com/bartovapps/gpstriprec/SplashScreen.java
//package com.bartovapps.gpstriprec;
//
//import android.content.Intent;
//import android.net.Uri;
//import android.os.AsyncTask;
//import android.os.Bundle;
//import android.os.Environment;
//import android.os.Handler;
//import android.os.Looper;
//import android.widget.ImageView;
//import android.widget.TextView;
//import android.widget.Toast;
//
//import androidx.appcompat.app.AppCompatActivity;
//
//import com.bartovapps.gpstriprec.presentation.screens.main_screen.MainScreen;
//import com.bartovapps.gpstriprec.utils.Utils;
//import com.squareup.picasso.Picasso;
//
//import java.io.BufferedInputStream;
//import java.io.FileNotFoundException;
//import java.io.FileOutputStream;
//import java.io.IOException;
//import java.io.InputStream;
//
//public class SplashScreen extends AppCompatActivity {
//
//    private static final String LOG_TAG = SplashScreen.class.getSimpleName();
//    private static final long SPLASH_TIMEOUT = 3500;
//
//    private String version;
//    private ImageView iv;
//    String root = Environment.getExternalStorageDirectory().toString();
//    String projectDir;
//    String attachmentFile;
//    Intent i;
//
//    @Override
//    protected void onCreate(Bundle savedInstanceState) {
//        super.onCreate(savedInstanceState);
//        setContentView(R.layout.splash_screen);
//        i = new Intent(SplashScreen.this, MainScreen.class);
//        attachmentFile = getExternalFilesDir(null) + "/" + "attachment.kml";
//        version = Utils.getApplicationVersion(SplashScreen.this);
//        iv = findViewById(R.id.splashImage);
//
//
//        Picasso.with(iv.getContext())
//                .load(R.drawable.splash)
//                .fit()
//                .centerInside()
//                .into(iv);
//
//        new Handler(Looper.getMainLooper()).postDelayed(this::startMainActivity, SPLASH_TIMEOUT);
//    }
//
//
//
//    private void startMainActivity() {
//        Intent launcherIntent = getIntent();
//        String path = null;
//
//        if (getIntent().getScheme() == null) {
//            startActivity(i);
//            finish();
//            return;
//        }
//
//        if (getIntent().getScheme().equals("file")) {
//            path = launcherIntent.getData().getPath();
////            Log.i(LOG_TAG, "Started from a local file..");
////            Log.i(LOG_TAG, "type is: " + launcherIntent.getType());
////            Log.i(LOG_TAG, "Host is: " + getIntent().getData().getHost());
//            if (path != null) {
//                i.putExtra("kml_path", path);
//            }
//            startActivity(i);
//            finish();
//            return;
//        }
//
//        if (getIntent().getScheme().equals("content")) {
////            Log.i(LOG_TAG, "Started from a email attachment..");
////            Log.i(LOG_TAG, "Type is: " + getIntent().getType());
////            Log.i(LOG_TAG, "Host is: " + getIntent().getData().getHost());
//            Uri uri = launcherIntent.getData();
//            if (uri != null) {
//                try {
//                    InputStream attachment = getContentResolver().openInputStream(uri);
//                    if (attachment == null) {
////                        Log.i(LOG_TAG, "not from mail attachment.. getting path from data..");
//                        startActivity(i);
//                    } else {
////                        Log.i(LOG_TAG, "Received file from attachment");
//                        UploadAttachmentTask uploadAttachmentTask = new UploadAttachmentTask();
//                        Container container = new Container();
//                        container.fileName = attachmentFile;
//                        container.uri = uri;
//                        uploadAttachmentTask.execute(container);
//                        //path = getFileFromAttachment(uri);
//                    }
////                    Log.i(LOG_TAG, "Got path: " + path);
//                } catch (FileNotFoundException e) {
//                    e.printStackTrace();
////                    Log.i(LOG_TAG, "File not found exception..");
//                } catch (SecurityException e){
//                    Toast.makeText(SplashScreen.this, getString(R.string.error_access_attachment), Toast.LENGTH_SHORT).show();
//                    startActivity(i);
//                    finish();
//                }
//            }
//        }
//
//
//
//    }
//
//
//
//
//    private class UploadAttachmentTask extends AsyncTask<Container, Void, Container> {
//
//        @Override
//        protected Container doInBackground(Container... params) {
//            Container container = params[0];
//
//            try {
//                InputStream attachment = getContentResolver().openInputStream(container.uri);
//                FileOutputStream tmp = new FileOutputStream(container.fileName);
//
//                BufferedInputStream bis = new BufferedInputStream(attachment);
//                StringBuffer b = new StringBuffer();
//
//                while (bis.available() != 0) {
//                    char c = (char) bis.read();
//                    b.append(c);
//                }
//
//                bis.close();
//                attachment.close();
//                tmp.write(String.valueOf(b).getBytes());
//                tmp.close();
//                container.status = true;
//            } catch (FileNotFoundException e) {
//                container.status = false;
////                Log.i(LOG_TAG, "File not found exception..");
//                e.printStackTrace();
//            } catch (IOException e) {
//                container.status = false;
////                Log.i(LOG_TAG, "IOException..");
//                e.printStackTrace();
//            }
//            return container;
//        }
//
//        @Override
//        protected void onPostExecute(Container container) {
////            Log.i(LOG_TAG, "file load status = " + container.status);
////            Log.i(LOG_TAG, "kml file name: " + container.fileName);
//            if(container.status == true){
//                if (container.fileName != null) {
//                    i.putExtra("kml_path", container.fileName);
//                }
//            }
//            startActivity(i);
//            finish();
//
//            return;
//        }
//
//        @Override
//        protected void onPreExecute() {
//        }
//
//        @Override
//        protected void onProgressUpdate(Void... values) {
//        }
//    }
//
//    class Container{
//        Uri uri;
//        String fileName;
//        boolean status;
//    }
//
//}

File: app/src/main/kotlin/presentation/screens/splash_screen/SplashScreen.kt
package com.bartovapps.gpstriprec.presentation.screens.splash_screen

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.AsyncTask
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.ImageView
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.bartovapps.gpstriprec.R
import com.bartovapps.gpstriprec.presentation.screens.main_screen.MainScreen
import com.bartovapps.gpstriprec.utils.Utils
import com.squareup.picasso.Picasso
import dagger.hilt.android.AndroidEntryPoint
import java.io.BufferedInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException


@AndroidEntryPoint
class SplashScreen : AppCompatActivity() {

    private val permissionsViewModel: PermissionsViewModel by viewModels()
    private var version: String? = null
    private lateinit var iv: ImageView
    var attachmentFile: String? = null

    private val requestPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { isGranted ->
        permissionsViewModel.onStepResult(isGranted)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.splash_screen)
        setupUi()
        setupObservers()

        permissionsViewModel.initPermissionChain(
            hasLocation = isGranted(Manifest.permission.ACCESS_FINE_LOCATION),
            hasNotifications = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
                isGranted(Manifest.permission.POST_NOTIFICATIONS) else true,
            hasMediaImages = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
                isGranted(Manifest.permission.READ_MEDIA_IMAGES) else true,
            hasStorage = if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q)
                isGranted(Manifest.permission.WRITE_EXTERNAL_STORAGE) else true
        )
    }


    private fun setupUi() {
        attachmentFile = getExternalFilesDir(null).toString() + "/" + "attachment.kml"
        version = Utils.getApplicationVersion(this@SplashScreen)
        iv = findViewById(R.id.splashImage)


        Picasso.with(this)
            .load(R.drawable.splash)
            .fit()
            .centerInside()
            .into(iv)
    }

    private fun isGranted(permission: String): Boolean {
        return ContextCompat.checkSelfPermission(this, permission) == PackageManager.PERMISSION_GRANTED
    }

    private fun setupObservers() {
        permissionsViewModel.currentStep.observe(this) { step ->
            if (step != null) {
                showRationaleDialog(step)
            } else if (permissionsViewModel.navigateToMain.value != true) {
                // If step is null and we aren't moving to main, a mandatory permission was denied
                finish()
            }
        }

        permissionsViewModel.navigateToMain.observe(this) { shouldNavigate ->
            if (shouldNavigate) {
                scheduleMainActivity()
            }
        }
    }


    private fun showRationaleDialog(step: PermissionsViewModel.PermissionStep) {
        AlertDialog.Builder(this)
            .setTitle(step.title)
            .setMessage(step.message)
            .setCancelable(false)
            .setPositiveButton("Allow") { _, _ ->
                requestPermissionLauncher.launch(step.permission)
            }
            .setNegativeButton("Refuse") { dialog, _ ->
                dialog.dismiss()
                permissionsViewModel.onStepResult(false)
            }
            .show()
    }
    private fun scheduleMainActivity() {
        Handler(Looper.getMainLooper()).postDelayed(
            Runnable { this.startMainActivity() },
            SPLASH_TIMEOUT
        )
    }

    private fun startMainActivity() {
        var path: String? = null

        if (intent.scheme == null) {
            startActivity(Intent(this@SplashScreen, MainScreen::class.java))
            finish()
            return
        }

        if (intent.scheme == "file") {
            path = intent.data?.path
            //            Log.i(LOG_TAG, "Started from a local file..");
//            Log.i(LOG_TAG, "type is: " + launcherIntent.getType());
//            Log.i(LOG_TAG, "Host is: " + getIntent().getData().getHost());
            val i = Intent(this@SplashScreen, MainScreen::class.java)
            path?.let {
                i.putExtra("kml_path", it)
            }
            startActivity(i)
            finish()
            return
        }


        if (intent.scheme == "content") {
//            Log.i(LOG_TAG, "Started from a email attachment..");
//            Log.i(LOG_TAG, "Type is: " + getIntent().getType());
//            Log.i(LOG_TAG, "Host is: " + getIntent().getData().getHost());
            val uri = intent.data
            uri?.let {
                try {
                    val attachment = contentResolver.openInputStream(uri)
                    if (attachment == null) {
//                        Log.i(LOG_TAG, "not from mail attachment.. getting path from data..");
                        val i = Intent(this@SplashScreen, MainScreen::class.java)
                        startActivity(i)
                    } else {
//                        Log.i(LOG_TAG, "Received file from attachment");
                        val uploadAttachmentTask = UploadAttachmentTask()
                        val container = Container()
                        container.fileName = attachmentFile
                        container.uri = uri
                        uploadAttachmentTask.execute(container)
                        //path = getFileFromAttachment(uri);
                    }
                    //                    Log.i(LOG_TAG, "Got path: " + path);
                } catch (e: FileNotFoundException) {
                    e.printStackTrace()
                    //                    Log.i(LOG_TAG, "File not found exception..");
                } catch (e: SecurityException) {
                    Toast.makeText(
                        this@SplashScreen,
                        getString(R.string.error_access_attachment),
                        Toast.LENGTH_SHORT
                    ).show()
                    val i = Intent(this@SplashScreen, MainScreen::class.java)
                    startActivity(i)
                    finish()
                }

            }
        }
    }


    private inner class UploadAttachmentTask : AsyncTask<Container, Void?, Container>() {
        override fun doInBackground(vararg params: Container): Container {
            val container: Container = params[0]

            try {

                val attachment = container.uri?.let {
                    contentResolver.openInputStream(it)
                }
                val tmp = FileOutputStream(container.fileName)

                val bis = BufferedInputStream(attachment)
                val b = StringBuffer()

                while (bis.available() != 0) {
                    val c = bis.read().toChar()
                    b.append(c)
                }

                bis.close()
                attachment?.close()
                tmp.write(b.toString().toByteArray())
                tmp.close()
                container.status = true
            } catch (e: FileNotFoundException) {
                container.status = false
                //                Log.i(LOG_TAG, "File not found exception..");
                e.printStackTrace()
            } catch (e: IOException) {
                container.status = false
                //                Log.i(LOG_TAG, "IOException..");
                e.printStackTrace()
            }
            return container
        }


        override fun onPostExecute(container: Container?) {
            if (container?.status == true) {
                if (container.fileName != null) {
                    val i = Intent(this@SplashScreen, MainScreen::class.java).also {
                        it.putExtra("kml_path", container.fileName)
                    }
                    startActivity(i)
                    finish()
                }
            }
        }

    }

    internal class Container {
        var uri: Uri? = null
        var fileName: String? = null
        var status: Boolean = false
    }

    companion object {
        private val LOG_TAG: String = SplashScreen::class.java.getSimpleName()
        private const val SPLASH_TIMEOUT: Long = 3500
    }
}